canal/canal.go

```diff
@@ -137,28 +142,59 @@ func (c *Canal) prepareDumper() error {
 	return nil
 }
 
-func (c *Canal) Start() error {
+func (c *Canal) Start(syncErrorCh chan error) error {
 	c.wg.Add(1)
-	go c.run()
+	go c.run(syncErrorCh)
 
 	return nil
 }
 
-func (c *Canal) run() error {
+func (c *Canal) run(syncErrorCh chan error) error {
 	defer c.wg.Done()
 
+	shouldStopRetrying := func(attemtp int, e error) bool {
+		if attemtp >= maxAttempts {
+			if e != nil {
+				// communicate the error to the upper layer (client)
+				// in case we exhaust all the attempts
+				syncErrorCh <- e
+				return true
+			}
+		}
+		return false
+	}
+
 	if err := c.tryDump(); err != nil {
 		log.Errorf("canal dump mysql err: %v", err)
 		return errors.Trace(err)
 	}
 
 	close(c.dumpDoneCh)
 
-	if err := c.startSyncBinlog(); err != nil {
+	attemptNum := 0
+	sleep := time.Duration(waitForRetrySeconds * time.Second)
+
+	for {
+		attemptNum++
+
+		errSync := c.startSyncBinlog();
+
 		if !c.isClosed() {
-			log.Errorf("canal start sync binlog err: %v", err)
+			log.Errorf("canal start sync binlog err: %v", errSync)
+		}
+
+		if shouldStopRetrying(attemptNum, errSync) {
+			return errSync
+		}
+
+		sleep = sleep * 2
+		time.Sleep(sleep)
+
+		errPrepare := c.prepareSyncer()
+
+		if shouldStopRetrying(attemptNum, errPrepare) {
+			return errPrepare
```

Comment:
Since this is going to be a long-running task, on success should we reset `attemptNum` and `sleep`?

If it fails 4 times and then succeeds and then a week or two later, it fails again, will it have another 5 attempts or just 1?